File: backend/app/models/friend.py
from enum import Enum
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, EmailStr, UUID4, validator, Field


class FriendshipStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    BLOCKED = "blocked"
    CANCELLED = "cancelled"


class UserProfile(BaseModel):
    user_id: UUID4
    email: EmailStr
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    is_public: bool = False
    created_at: datetime
    updated_at: datetime


class UserProfileUpdate(BaseModel):
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    is_public: Optional[bool] = None

    @validator("display_name")
    def validate_display_name(cls, v):
        if v is not None:
            if len(v.strip()) == 0:
                raise ValueError("Display name cannot be empty")
            if len(v) > 50:
                raise ValueError("Display name must be 50 characters or less")
        return v


class Friendship(BaseModel):
    # id: UUID4
    user_one_id: UUID4
    user_two_id: UUID4
    action_user_id: UUID4
    status: FriendshipStatus
    created_at: datetime
    updated_at: datetime


class FriendRequest(BaseModel):
    email: EmailStr


class LeaderboardEntry(BaseModel):
    rank: int
    user_id: UUID4
    display_name: Optional[str] = None
    email: str
    total_experience: int = Field(default=0)
    level: int = Field(default=1)
    tasks_completed: int = Field(default=0)
    plants_grown: int = Field(default=0)
    longest_streak: int = Field(default=0)
    current_streak: int = Field(default=0)

File: frontend/src/types/friend.ts
export type FriendshipStatus = "pending" | "accepted" | "declined" | "blocked";

export interface UserProfile {
  user_id: string;
  email: string;
  display_name?: string;
  avatar_url?: string;
  is_public: boolean;
  created_at: string;
  updated_at: string;
}

export interface Friendship {
  user_one_id: string;
  user_two_id: string;
  action_user_id: string;
  status: FriendshipStatus;
  created_at: string;
  updated_at: string;
  profile?: UserProfile; // Joined from user_profiles
}

export interface FriendRequest {
  email: string;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  display_name?: string;
  email: string;
  total_experience: number;
  level: number;
  tasks_completed: number;
  plants_grown: number;
  longest_streak: number;
  current_streak: number;
}

export interface UserProfileUpdate {
  display_name?: string;
  avatar_url?: string;
  is_public?: boolean;
}
